File: bfs/JunYoungKr/1012.py
from collections import deque
import sys
 
input = sys.stdin.readline

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def bfs(x, y):
    queue = deque([(x, y)])
    
    # 방문한 좌표는 2로 바꾸어 재방문하지 않도록 처리
    graph[x][y] = 2
    
    # 큐가 빌 때까지
    while queue:
        x, y = queue.popleft()
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            # 범위를 벗어나면 무시
            if nx < 0 or nx >= N or ny < 0 or ny >= M:
                continue
            # 배추가 있는 곳이면 방문 처리하고 큐에 추가
            if graph[nx][ny] == 1:
                graph[nx][ny] = 2
                queue.append((nx, ny))
                
    return count          
        
        
# 테스트 케이스 T 입력
T = int(input())

for _ in range(T):
    # 배추밭의 가로길이 M, 세로길이 N, 배추가 심어져 있는 위치의 개수 K 입력
    M, N, K = map(int, input().split())
    
    # 가로 M, 세로 N 크기의 배추밭 배열 생성
    graph = [[0] * M for _ in range(N)]
    
    # K줄 배추의 위치 X, Y 입력 / 각각의 범위는 0 <= X <= M-1, 0 <= Y <= N-1
    for _ in range(K):
        x, y = map(int, input().split())
        graph[y][x] = 1
        
    count = 0
    for i in range(N):
        for j in range(M):
            if graph[i][j] == 1:  # 배추가 심어진 곳을 발견하면
                bfs(i, j)  # 그 구역을 탐색
                count += 1  # 구역을 하나 찾았으므로 count 증가
    print(count)
    

File: BFS/DonggyunKim00/1012.js
const fs = require('fs');

const filePath = process.platform === 'linux' ? '/dev/stdin' : 'input.txt';
const input = fs
  .readFileSync(filePath, 'utf8')
  .toString()
  .trim()
  .split('\n')
  .map((item) => item.split(' ').map(Number));

const [dx, dy] = [
  [1, -1, 0, 0],
  [0, 0, 1, -1],
];

const bfs = (info, location) => {
  // 변수 선언
  const [M, N, K] = info;
  let visited = Array.from({ length: M }, () =>
    Array.from({ length: N }, () => 0)
  );

  // 그래프 만들기
  let graph = Array.from({ length: M }, () =>
    Array.from({ length: N }, () => 0)
  );
  for (let i = 0; i < K; i++) {
    const [x, y] = location[i];
    graph[x][y] = 1;
  }

  // location 배열을 모두 돌면서 각영역에 대한 bfs 로직을 한번씩 돌리는데
  // 이때 각 영역을 한번 돌때마다 count를 1씩 증가
  let count = 0;
  for (let i = 0; i < location.length; i++) {
    const [x, y] = location[i];
    if (visited[x][y]) continue;

    // bfs 로직
    const queue = [];

    let curr = [x, y];
    queue.push(curr);
    visited[curr[0]][curr[1]] = 1;

    while (queue.length) {
      const [curr_x, curr_y] = queue.shift();

      for (let dir = 0; dir < 4; dir++) {
        const move_x = curr_x + dx[dir];
        const move_y = curr_y + dy[dir];

        if (move_x < 0 || move_x >= M || move_y < 0 || move_y >= N) continue;
        if (!graph[move_x][move_y] || visited[move_x][move_y]) continue;

        queue.push([move_x, move_y]);
        visited[move_x][move_y] = 1;
        curr = [move_x, move_y];
      }
    }
    count += 1;
  }

  return count;
};

const solution = (T, inputs) => {
  let index = 0;

  const answer = [];
  for (let i = 0; i < T; i++) {
    const [M, N, K] = inputs[index];
    const arr = inputs.slice(index + 1, index + K + 1);

    answer.push(bfs([M, N, K], arr));

    index += K + 1;
  }

  return answer;
};

const [T, ...inputs] = input;
const answer = solution(T[0], inputs).join('\n');
console.log(answer);
